### ISSUE TITLE
[arithmetic_side_effects] false positive with string and derefs

### ISSUE THREAD

**Issue by JP-Ellis** [rust-lang/rust-clippy#14054]
### Summary

Concatenation of a `String` with `&str` when the `&str` is a `String` deref produces a false positive.

### Lint Name

arithmetic_side_effects

### Reproducer

Here's a minimal example that produces the error:

```rust
let x = String::new();
let y = String::new();
let z = x + &y;

assert_eq!(&z, "");
```

[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=4425ed121c10a4e4b6750816eec2b55e)

### Version

```text
rustc 1.84.0 (9fc6b4312 2025-01-07)
binary: rustc
commit-hash: 9fc6b43126469e3858e2fe86cafb4f0fd5068869
commit-date: 2025-01-07
host: aarch64-apple-darwin
release: 1.84.0
LLVM version: 19.1.5
```

(also reproduced in Rust playground)

**Comment by JP-Ellis** [comment]
Thanks for the super speedy fix!

**Comment by Iizuki** [comment]
Has this regressed? Playground reports the clippy warning (again?) on 1.92.0.

**Comment by samueltardieu** [comment]
It looks like it might not have been entirely fixed. I'll have a look at this.

@rustbot claim
